packages/frontend/editor-ui/src/utils/nodeViewUtils.ts

```diff
@@ -48,12 +56,68 @@ export const getLeftmostTopNode = <T extends { position: XYPosition }>(nodes: T[
 	}, nodes[0]);
 };
 
+export const getLeftMostNode = <T extends { position: XYPosition }>(nodes: T[]): T => {
+	return nodes.reduce((leftmost, node) => {
+		if (node.position[0] < leftmost.position[0]) {
+			return node;
+		}
+
+		return leftmost;
+	}, nodes[0]);
+};
+
+export const getTopMostNode = <T extends { position: XYPosition }>(nodes: T[]): T => {
+	return nodes.reduce((topmost, node) => {
+		if (node.position[1] < topmost.position[1]) {
+			return node;
+		}
+
+		return topmost;
+	}, nodes[0]);
+};
+
+export const getRightMostNode = <T extends { position: XYPosition }>(nodes: T[]): T => {
+	return nodes.reduce((rightmost, node) => {
+		if (node.position[0] > rightmost.position[0]) {
+			return node;
+		}
+
+		return rightmost;
+	}, nodes[0]);
+};
+
+export const getBottomMostNode = <T extends { position: XYPosition }>(nodes: T[]): T => {
+	return nodes.reduce((bottommost, node) => {
+		if (node.position[1] > bottommost.position[1]) {
+			return node;
+		}
+
+		return bottommost;
+	}, nodes[0]);
+};
+
+export const getNodesGroupSize = (nodes: INodeUi[]): [number, number] => {
```

Comment:
There is some similar code in useCanvasLayout: `compositeBoundingBox`. It works from VueFlow GraphNode so it also has access to node width/height

Reply:
I will probably not be able to use `compositeBoundingBox` because that would tie the node addition operation to VueFlow, and everything in `useCanvasOperations` right now is independent from the UI, mainly bound to the store.